File: fsharp/src/JsonTag.FSharp/JsonTagOptions.fs
module JsonTag.FSharp.JsonTagOptions

open System.Text.Json
open System.Text.Json.Serialization // JsonFSharpOptions, JsonUnionEncoding from FSharp.SystemTextJson

/// Applies the JSON# convention to existing JsonFSharpOptions.
/// Sets unionTagName to "#type" and ensures InternalTag + NamedFields encoding.
/// All other settings from the source options are preserved.
let apply (source: JsonFSharpOptions) =
    JsonFSharpOptions(
        unionEncoding = (source.UnionEncoding ||| JsonUnionEncoding.InternalTag ||| JsonUnionEncoding.NamedFields),
        unionTagName = "#type",
        unionFieldsName = source.UnionFieldsName,
        unionTagNamingPolicy = source.UnionTagNamingPolicy,
        unionFieldNamingPolicy = source.UnionFieldNamingPolicy,
        unionTagCaseInsensitive = source.UnionTagCaseInsensitive,
        allowNullFields = source.AllowNullFields,
        includeRecordProperties = source.IncludeRecordProperties,
        types = source.Types,
        allowOverride = source.AllowOverride
    )

/// Pre-configured JsonSerializerOptions with #type as the discriminator,
/// camelCase property names, and compact output.
let Default =
    let fSharpOptions =
        JsonFSharpOptions(
            unionEncoding = (JsonUnionEncoding.InternalTag ||| JsonUnionEncoding.NamedFields ||| JsonUnionEncoding.UnwrapOption ||| JsonUnionEncoding.UnwrapSingleCaseUnions ||| JsonUnionEncoding.UnwrapFieldlessTags),
            unionTagName = "#type"
        )
    let options = JsonSerializerOptions(PropertyNamingPolicy = JsonNamingPolicy.CamelCase, WriteIndented = false, AllowOutOfOrderMetadataProperties = true)
    fSharpOptions.AddToJsonSerializerOptions(options)
    options

File: csharp/src/JsonTag/JsonTagOptions.cs
using System.Text.Json;

namespace JsonTag;

public static class JsonTagOptions
{
    public static JsonSerializerOptions Default { get; } = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        AllowOutOfOrderMetadataProperties = true,
        WriteIndented = false,
        TypeInfoResolver = new JsonTagResolver(),
    };

    /// <summary>
    /// Returns a copy of the given options with JsonTag support added.
    /// </summary>
    /// <remarks>
    /// This replaces any existing TypeInfoResolver on the options.
    /// </remarks>
    public static JsonSerializerOptions WithJsonTag(this JsonSerializerOptions options) =>
        new(options)
        {
            AllowOutOfOrderMetadataProperties = true,
            TypeInfoResolver = new JsonTagResolver(),
        };
}
